ios/LayoutReanimation/REASharedTransitionManager.m

```diff
@@ -0,0 +1,518 @@
+      }
+    }
+  }
+  while (view != nil && ![view isKindOfClass:[RNSScreenStackView class]] && view.superview != nil) {
+    view = view.superview;
+  }
+  if ([view isKindOfClass:[RNSScreenStackView class]]) {
+    return view;
+  }
+#endif
+  return nil;
+}
+
+- (void)observeChanges:(UIView *)view
+{
+  if (view == nil || [_screenHasObserver containsObject:view.reactTag]) {
+    return;
+  }
+  [_screenHasObserver addObject:view.reactTag];
+
+  [view addObserver:self forKeyPath:@"transitionDuration" options:NSKeyValueObservingOptionNew context:nil];
+  [view addObserver:self forKeyPath:@"activityState" options:NSKeyValueObservingOptionNew context:nil];
+
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    // it replaces method for RNSScreenView class, so it can be done only once
+    UIViewController *viewController = [view valueForKey:@"controller"];
+    [self swizzleMethod:@selector(viewDidLayoutSubviews)
+                   with:@selector(swizzled_viewDidLayoutSubviews)
+               forClass:[viewController class]];
+    [self swizzleMethod:@selector(notifyWillDisappear)
+                   with:@selector(swizzled_notifyWillDisappear)
+               forClass:[view class]];
+  });
+}
+
+- (void)swizzleMethod:(SEL)originalSelector with:(SEL)swizzledSelector forClass:(Class)originalClass
+{
+  Class class = [self class];
+  Method originalMethod = class_getInstanceMethod(originalClass, originalSelector);
+  Method swizzledMethod = class_getInstanceMethod(class, swizzledSelector);
+  IMP originalImp = method_getImplementation(originalMethod);
+  IMP swizzledImp = method_getImplementation(swizzledMethod);
+  class_replaceMethod(originalClass, swizzledSelector, originalImp, method_getTypeEncoding(originalMethod));
+  class_replaceMethod(originalClass, originalSelector, swizzledImp, method_getTypeEncoding(swizzledMethod));
+}
+
+- (void)swizzled_viewDidLayoutSubviews
+{
+  /*
+    The screen header is added later than other screen children and changes their
+    position on the screen. That's why we need to make a snapshot after the header
+    mount. We don't want to add a strong dependency to the react-native-screens
+    library so we decided to use KVO from Obj-C. However, all properties of the
+    screen are updated without calling Obj-C's setters. We swizzled the method
+    `viewDidLayoutSubviews` - this method is called after the appearance of the header.
+    Then we call KVO to notify Reanimated observer about the attached header.
+  */
+
+  // call original method from react-native-screens, self == RNScreen
+  [self swizzled_viewDidLayoutSubviews];
+  // get RNScreenView from RNScreen to call KVO on RNScreenView because
+  // RNScreen doesn't contains writable public property.
+  UIView *view = [self valueForKey:@"screenView"];
+  // call KVO to run runAsyncSharedTransition from Reanimated
+  [view setValue:[view valueForKey:@"transitionDuration"] forKey:@"transitionDuration"];
```

Comment:
If we already do method swizzling why can't we directly call that method (in this case `runAsyncSharedTransition`) from here instead of adding KVO in between?

Reply:
That was because swizzled method changes the context of execution, so the `self` in swizzled method is `RNSScreen` instead of `REASharedTransitionManager`. That's why I used KVO to change the context of execution back to `REASharedTransitionManager`. But just now I found much better solution. I'll add it to #4014 soon.

Reply:
Resolved by: https://github.com/software-mansion/react-native-reanimated/pull/4014/commits/2110fb5c5762b19cf5555a0c34696a7c8836f7d1